editor/project_manager/project_list.cpp

```diff
@@ -486,7 +488,13 @@ ProjectList::Item ProjectList::load_project_data(const String &p_path, bool p_fa
 		ProjectManager::get_singleton()->add_new_tag(tag);
 	}
 
-	return Item(project_name, description, project_version, tags, p_path, icon, main_scene, unsupported_features, last_edited, p_favorite, grayed, missing, config_version);
+	// We can't use globalize_path because there is no loaded project. So we replicate the behavior of accessing user://
+	if (!cf_project_name.is_empty()) {
+		String safe_mode_lock_file = OS::get_singleton()->get_user_data_dir(cf_project_name).path_join(".safe_mode_lock");
```

Comment:
> Shouldn't it do get_safe_dir_name on a name, like it's done in the actualt project,

Yup, I missed that. I also don't remember why I used the raw `cf_project_name` variable instead of the unescaped `project_name`, so I'll revisit this.

> also seems like it won't handle application/config/use_custom_user_dir/application/config/custom_user_dir_name override

This is indeed a problem, because this query is inside the `get_user_dir`. I already had to move the query for `application/config/name` outside to be able to set it from the context of the project manager. My best idea so far is to extract all project specific logic from this function, and move this logic to the generic `os.cpp::get_user_data_dir()` function, with the OS-specific overrides now becoming quite simple:

```cpp
// os.cpp
String OS::get_user_data_dir() const {
	String appname = get_safe_dir_name(GLOBAL_GET("application/config/name"));
	if (!appname.is_empty()) {
		bool use_custom_dir = GLOBAL_GET("application/config/use_custom_user_dir");
		if (use_custom_dir) {
			String custom_dir = get_safe_dir_name(GLOBAL_GET("application/config/custom_user_dir_name"), true);
			if (custom_dir.is_empty()) {
				custom_dir = appname;
			}
			return get_user_data_dir(custom_dir);
		} else {
			return get_user_data_dir(get_godot_dir_name().path_join("app_userdata").path_join(appname));
		}
	}

	return get_user_data_dir(get_godot_dir_name().path_join("app_userdata").path_join("[unnamed project]"));
}
...

// os_unix.cpp
String OS_Unix::get_user_data_dir(const String &p_dirname) const {
	return get_data_path().path_join(p_dirname);
}
...

// os_windows.cpp
String OS_Unix::get_user_data_dir(const String &p_dirname) const {
	return get_data_path().path_join(p_dirname).replace("\\", "/");
}
``` 
What do you think?

Reply:
I ended up extracting this logic to `os.cpp`, which has the benefit of unifying 3 implementations (Unix, Windows and Web).